app/assets/javascripts/components/vm_cloud/vm-cloud-add-security-group-component.js

```diff
@@ -0,0 +1,61 @@
+ManageIQ.angular.app.component('vmCloudAddSecurityGroupComponent', {
+  bindings: {
+    recordId: '@',
+  },
+  controllerAs: 'vm',
+  controller: vmCloudAddSecurityGroupFormController,
+  templateUrl: '/static/vm_cloud/add_security_group.html.haml',
+});
+
+vmCloudAddSecurityGroupFormController.$inject = ['API', 'miqService'];
+
+function vmCloudAddSecurityGroupFormController(API, miqService) {
+  var vm = this;
+
+  vm.$onInit = function() {
+    vm.afterGet = false;
+    vm.vmCloudModel = {
+      security_group: null,
+    };
+    vm.security_groups = [];
+    vm.formId = vm.recordId;
+    vm.model = "vmCloudModel";
+    vm.saveable = miqService.saveable;
+    miqService.sparkleOn();
+    API.get("/api/vms/" + vm.recordId).then(function(data) {
```

Comment:
Aah.. the nice thing about promises is that you don't have to nest them, you can always return a Promise instead...

This part should be:

```js
var tenantId;
var currentSecurityGroups;

API.get("/api/vms/" + vm.recordId)
  .then(function(data) {
    tenantId = data.cloud_tenant_id;

    return API.get("/api/vms/" + vm.recordId + "/security_groups?expand=resources&attributes=id,name");
  })
  .then(function(data) {
    currentSecurityGroups = data.resources;

    return API.get("/api/cloud_tenants/" + tenantId + "/security_groups?expand=resources&attributes=id,name");
  })
  .then(function(data) {
    vm.security_groups = data.resources.filter(function(securityGroup) {
      return ! _.find(currentSecurityGroups, { id: securityGroup.id });
    });

    vm.afterGet = true;
    vm.modelCopy = angular.copy(vm.vmCloudModel);
  })
  .catch(miqService.handleFailure);
```

Reply:
..and since the first 2 queries don't depend on each other, they can run in parallel..

```js
var currentSecurityGroups;

$q.all([
  API.get("/api/vms/" + vm.recordId),
  API.get("/api/vms/" + vm.recordId + "/security_groups?expand=resources&attributes=id,name"),
])
  .then(function(data) {
    var tenantId = data[0].cloud_tenant_id;
    currentSecurityGroups = data[1].resources;

    return API.get("/api/cloud_tenants/" + tenantId + "/security_groups?expand=resources&attributes=id,name");
  })
  .then(function(data) {
    vm.security_groups = data.resources.filter(function(securityGroup) {
      return ! _.find(currentSecurityGroups, { id: securityGroup.id });
    });

    vm.afterGet = true;
    vm.modelCopy = angular.copy(vm.vmCloudModel);
  })
  .catch(miqService.handleFailure);
```